### DISCUSSION TITLE
[Improvement] Export collections using Postman format

### DISCUSSION THREAD

**Discussion by KevinSnyderCodes** [Kong/insomnia#3703] [Ideas]
Insomnia supports exporting collections in the Insomnia and HAR formats. I would like to add the ability to export using the Postman collection format(s).

I have spent a lot of time building an extensive collection in Insomnia that would be of great use to others in my organization, but most of them use Postman, and currently there is no way to export collections to a Postman-compatible format. I have no intention of making it more attractive for people to use Postman (quite the opposite in fact), but this would be a useful feature that would allow teams to use the REST client of their choice while still being able to share collections.

Documentation for the various Postman collection JSON schemas can be found at https://schema.getpostman.com

If this is something you think would be useful, I would be happy to implement it in my spare time and open a PR.

**Comment by gschier** [comment]
I'm completely okay with this being a community-supported feature of Insomnia and Postman is definitely popular enough to justify including it in Insomnia.

Let me know if you have any questions on how to get started 👍

**Comment by qkdreyer** [comment]
@KevinSnyderCodes any updates on this ?

**Comment by jacobhummel** [comment]
Would also love this. Did you end up working on it @KevinSnyderCodes?

**Comment by toclean** [comment]
Was also looking into adding this. Have you made any progress @KevinSnyderCodes?

**Comment by dmhalejr** [comment]
I've got a working prototype of at least GraphQL calls being ported properly into Postman collections v2.1.0 schema. @KevinSnyderCodes are you still working on this? Is there any extended post call examples you can provide @gschier? 

Will probably just use it for my purposes for now and strengthen the code if I should make it feature candidate quality

**Comment by ThaDaVos** [comment]
What's the ETA of this? I prefer using Insomnia over Postman, but some colleagues rather use Postman

**Comment by yaser-ali-s** [comment]
Just checking in again one year later. Any progress on this? What can we do to help?

**Comment by dmhalejr** [comment]
During the course of developing this enhancement, Postman began supporting GraphQL functionality out of the box which fit my use case for me.

https://learning.postman.com/docs/sending-requests/supported-api-frameworks/graphql/

**Comment by yaser-ali-s** [comment]
Ok so I don't mind contributing since this has gone stale. The purpose of the feature was to allow exporting so those who use postman can import accordingly; namely, colleagues. If there is anything who can pass on their work so I may continue or someone may point me to the right direction to start I'd appreciate it. Otherwise I'll try to do this ASAP; preferably by this weekend. Nudge me if I forget.

**Comment by Vyoam** [comment]
If you're ok with sharing your data to third party servers, https://www.apimatic.io/transformer/#supported-formats can do the Insomnia to Postman conversion.

For offline conversion, I have created a rudimentary tool for basic cases: https://github.com/Vyoam/InsomniaToPostmanFormat

**Comment by amcsi** [comment]
@yaser-ali-s here's your reminder 🙏

**Comment by jiili** [comment]
@yaser-ali-s this would be awesome 👍

**Comment by yaharga** [comment]
So far I figured out how to start:

- In the file *[packages/insomnia-app/app/ui/redux/modules/global.js](https://github.com/yaharga/insomnia/blob/export-to-postman/packages/insomnia-app/app/ui/redux/modules/global.js#L404)* I edited so that it would display "Postman" in the export dropdown (hopefully I'll export to both versions, but I'll start with V2.0).
- I then added the [if-statement](https://github.com/yaharga/insomnia/blob/export-to-postman/packages/insomnia-app/app/ui/redux/modules/global.js#L504) that would prepare for an additional export method.

Now for the process of conversion, I have one of three options:
- Go to the [Postman schema](https://schema.getpostman.com/) and create the conversion module manually.
- Use the [Insomnia Importers package](https://github.com/Kong/insomnia/tree/develop/packages/insomnia-importers/src/importers) to reverse-engineer what the Postman importer does to export.
- Use the [Har2Postman](https://github.com/javieraviles/Har2Postman) package to do the conversion (Insomnia>HAR>Postman).
- Use the [JS file](https://github.com/Vyoam/InsomniaToPostmanFormat) @Vyoam developed.

Ideally, I would like to go with the first option; however, I'm not sure I understand what I'm looking at when I look at the JSON schema, as in where to start with the conversion.

Just a progress update and if anyone has any inputs you're more than welcome. I'm sure I'll figure it out once I get more time to read into the schemas.

**Comment by Vyoam** [comment]
> So far I figured out how to start:
> 
> * In the file _[packages/insomnia-app/app/ui/redux/modules/global.js](https://github.com/yaharga/insomnia/blob/export-to-postman/packages/insomnia-app/app/ui/redux/modules/global.js#L404)_ I edited so that it would display "Postman" in the export dropdown (hopefully I'll export to both versions, but I'll start with V2.0).
> * I then added the [if-statement](https://github.com/yaharga/insomnia/blob/export-to-postman/packages/insomnia-app/app/ui/redux/modules/global.js#L504) that would prepare for an additional export method.
> 
> Now for the process of conversion, I have one of three options:
> 
> * Go to the [Postman schema](https://schema.getpostman.com/) and create the conversion module manually.
> * Use the [Insomnia Importers package](https://github.com/Kong/insomnia/tree/develop/packages/insomnia-importers/src/importers) to reverse-engineer what the Postman importer does to export.
> * Use the [Har2Postman](https://github.com/javieraviles/Har2Postman) package to do the conversion (Insomnia>HAR>Postman).
> * Use the [JS file](https://github.com/Vyoam/InsomniaToPostmanFormat) @Vyoam developed.
> 
> Ideally, I would like to go with the first option; however, I'm not sure I understand what I'm looking at when I look at the JSON schema, as in where to start with the conversion.
> 
> Just a progress update and if anyone has any inputs you're more than welcome. I'm sure I'll figure it out once I get more time to read into the schemas.

Thanks Yaser. Minor inputs from me...
* I think option 3 won't be good, as har doesn't support folder structure AFAIK.
* I think option 1 would be good, along with leveraging option 2 (learning from Postman importer flow). Leveraging option 4 (My converter) can be definitely useful - it handles folder structure and basic POST requests. But as you might have notices, some major and many minor cases are missing in my code. Also, that converter expects exported Insomnia json. I'm not sure how that data is represented internally in Insomnia.

**Comment by mxmlnglt** [comment]
The thing is, we should get some advice from the (current) official developers, because as for a time now, @gschier is not working on Insomnia anymore.

**Comment by dimitropoulos** [comment]
I'm a developer on the insomnia team and I agree with Greg's initial statement/sentiment.  I don't make product decisions so take that with a grain of salt.  I can say that my experience on the team so far has been that we wouldn't see some kind of competitive disadvantage to having a feature like this.  If someone is interested in submitting a PR, I'd be happy to take a peek.  At the same time, I am not aware of plans for our team to implement this.  We're pretty fully utilized with some new (and exciting!) things.

I also just recently spent a bunch of time with the importers package so I'm at least very familiar with how the interop between our format and postman's format works.

**Comment by TahirJalilov** [comment]
I need it too

**Comment by Vyoam** [comment]
For anyone interested, the standalone converter script ([InsomniaToPostmanFormat](https://github.com/Vyoam/InsomniaToPostmanFormat), also discussed above) now covers most of the major use cases.

**Comment by brunopulis** [comment]
> For anyone interested, the standalone converter script ([InsomniaToPostmanFormat](https://github.com/Vyoam/InsomniaToPostmanFormat), also discussed above) now covers most of the major use cases.

Thanks for this! Works perfectly.

**Comment by Vyoam** [comment]
I have raised the PR to integrate my aforementioned script functionality to Insomnia: https://github.com/Kong/insomnia/pull/3827/files
P.S. As visible in the PR, the Insomnia team has recommended using the plugin approach. Will try to start soon.

Limitations:
* These elements/data aren't being handled: Auth, Environment, Cookies, Docs, YAML requests, EDN requests, Graph Query requests.
* With respect to query params, Insomnia supports them to be added directly in the URL, but also through the 'Query' form. If such 'mixed' input is present, the output would erroneously have first two params prefixed with question marks instead of the just the first param.
* I think the the listing order of the various requests/folders isn't guaranteed to be same as that of the input.

**Comment by Vyoam** [comment]
Beta: https://github.com/Vyoam/insomnia-plugin-postman-export
